docs(modal): usage tab updates (#726)
carbon-design-system/carbon-website: src/pages/components/modal/usage.mdx

</Column>
</Row>

The modal is composed of three distinct zones: A header, the body, and a footer. Components (eg. data table, form, progress indicator) can occupy the full width of the modal.

1. **Header**: Includes a title, optional label, and the close icon. The title should be brief and clearly describe the dialogs’s task or purpose. Use the optional label above the title set the context for the information in the dialog.

2. **Body**: Contains the information and/or controls needed to complete the modal's task. It can include message text and components. A modal should include only content relevant to completing the current task. Text should only be 80% of the modal’s width and components can span 100% of the width.

3. **Footer**: Contains the main actions needed to complete or cancel the dialog task. Button groupings change based on modal variant. Use descriptive words for the actions like Add, Delete, Save and avoid vague words like Done or OK. If you need to include a “docs” or other non-primary action, include it as a link in the modal’s body.

4. **x**: The close `x` icon will close the dialog without submitting any data.

> 5. **Overlay**: Screen overlay that obscures the on page content.

Reviewer: ```suggestion
5. **Overlay**: Screen overlay that obscures the on-page content.
```

## Modal sizes

There are four preset modal sizes: **xs, small, default, large**. They have responsive widths that change based on the screen size. Modal widths are defined as screen percentages that will complement columns on the 2x grid. See the modal spec for specifics.

Choose a size best for the amount of modal content you have. Modals with short messages should use a xs or small modal to avoid long single lines; for complex components, like data table the default or large modal will be more accommodating.

<Row>
<Column colLg={8}>

![Modal sizes](images/modal-usage-sizes.png)

</Column>
</Row>

#### Title as message

> For short, direct messages the title can include the whole message to add visual clairty to an otherwise repeativite title and body message. When using this style,no other body copy may be included.

Reviewer: ```suggestion
For short, direct messages the title can include the whole message to add visual clarity to an otherwise repetitive title and body message. When using this style, no other body copy may be included.
```

<Row>
<Column colLg={8}>

![Modal sizes](images/modal-usage-title.png)

</Column>
</Row>

### Max heights

Each modal size has a max height in order to maintain a proper window ratio. If your modal has too much scrolling due to a max height consider using the next modal size up. If the large modal height is still not enough space then this is an indicator that a full page may be needed instead.

<!--
| Modal size | Max height |
| ---------- | ---------- |
| XS         | 48% |
| Small | 72% |
| Default | 84% |
| Large | 96% |
-->

<Row>
<Column colLg={8}>

![Modal sizes](images/modal-usage-max-heights.png)

</Column>
</Row>

#### Scrolling behavior

> When the modal content is longer than the modal height then the body section should scroll vertically with the header and footer remaining fixed in place. The content should visible fade at the end of the modal body area to indicate there is additional content out of view. Modal content should never horizontal scroll, instead use a larger size modal.

Reviewer: ```suggestion
When the modal content is longer than the modal height then the body section should scroll vertically with the header and footer remaining fixed in place. The content should visibly fade at the end of the modal body area to indicate there is additional content out of view. Modal content should never scroll horizontally; instead, use a larger size modal.
```

## Variations

There are four main variations of modal.

| Variant                                                        | Usage                                                                                                                           |
| -------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- |
| [Passive](/components/modal/usage#passive-modal)               | Presents information the user needs to be aware of concerning their current workflow. Contains no actions for the user to take. |
| [Transactional](/components/modal/usage#transactional-modal)   | Requires an action to be taken in order for the modal to be completed and closed. Contains a cancel and primary action buttons. |
| [Acknowledgment](/components/modal/usage#acknowledgment-modal) | System requires an acknowledgement of the information from the user. Contains only a single button, commonly "OK."              |

> | [Progress](/components/modal/usage#progress-modal)             | Requires several steps to be completed before it can be closed. Contains a cancel, previous and next/completion buttons.        |

Reviewer: ```suggestion
| [Progress](/components/modal/usage#progress-modal)             | Requires several steps to be completed before it can be closed. Contains a cancel, previous, and next/completion buttons.        |
```

- **Click elsewhere**: Clicking outside the passive modal area will automatically close the modal.
- **Esc**: Press `ESC` on the keyboard

<Row>
<Column colLg={8}>

![Passive modal example](images/modal-usage-passive.png)

</Column>
</Row>

### Transactional modal

Transactional modals are used to validate user decisions or to gain secondary confirmation from the user. Transactional modals require an action to be taken in order for the modal to be completed and closed. It contains a cancel and primary action buttons.

#### Dismissing a transactional modal

Transactional modals are persistent until dismissed in one of the following ways.

- **Task completion**: clicking the primary action will complete the task and automatically close the modal.
- **Cancel button**: clicking the cancel button will close the modal and return the user to its previous context. Cancel undoes all applied changes.
- **x**: Clicking the close `x` icon in the upper right will close the modal without submitting any data and return the user to its previous context.

> - **Esc**: Press `ESC` on the keyboard

Reviewer: ```suggestion
- **Esc**: Press `ESC` on the keyboard.
```

<Row>
<Column colLg={8}>

![Three button modal example](images/modal-usage-three-buttons.png)

</Column>
</Row>

#### Danger modal

Danger modal is a specific type of transactional modal used for destructive or irreversible actions. In danger modal, the primary button is replaced by a danger button. They are commonly used in high impact moments as a confirmation for an action that would result in a significant data loss if done accidentally.

<Row>
<Column colLg={8}>

![danger modal](images/modal-usage-danger.png)

</Column>
</Row>

### Acknowledgment modal

Acknowledgment modals are used when the system requires the user to acknowledge the presented information. It contains only a single button, commonly “OK.”  Acknowledgment is confirmed when the user clicks the primary button.

#### Dismissing an acknowledgment modal

Acknowledgment modals are persistent until dismissed in one of the following ways.

> - **Task completion**: clicking the primary action will complete the task and automatically close the modal.

Reviewer: ```suggestion
- **Task completion**: Clicking the primary action will complete the task and automatically close the modal.
```

> - **x**: Clicking the close `x` icon in the upper right will close the modal without submitting the user's acknowledgement and return the user to its previous context.

Reviewer: ```suggestion
- **x**: Clicking the close `x` icon in the upper right will close the modal without submitting the user's acknowledgement and return the user to the previous context.
```

> - **Esc**: Press `ESC` on the keyboard

Reviewer: ```suggestion
- **Esc**: Press `ESC` on the keyboard.
```

#### Single button

Single buttons are placed on the right side, span 50% of the modal, and bleed to the edge. In most scenarios, a primary button is used when one button is needed.

<Row>
<Column colLg={8}>

![One button modal example](images/modal-usage-one-button.png)

</Column>
</Row>

### Progress modal

For longer tasks, use a progress modal to give the user a sense of completion and orientation within the focused flow. For complex flows with complex choices, consider using a full page instead of a modal. A progress modal is not a solution for excess modal content. It should only be used to present information in more consumable and focused chunks.

#### Dismissing a progress modal

Transactional modals are persistent until dismissed in one of the following ways.

> - **Task completion**: clicking the primary action will complete the task and automatically close the modal.

Reviewer: ```suggestion
- **Task completion**: Clicking the primary action will complete the task and automatically close the modal.
```

> - **Cancel button**: clicking the cancel button will close the modal and return the user to its previous context. Cancel undoes all applied changes.

Reviewer: ```suggestion
- **Cancel button**: Clicking the Cancel button will close the modal and return the user to its previous context. Cancel undoes all applied changes.
```

- **x**: Clicking the close `x` icon in the upper right will close the modal without submitting any data and return the user to its previous context.

> - **Esc**: Press `ESC` on the keyboard

Reviewer: ```suggestion
- **Esc**: Press `ESC` on the keyboard.
```
